### PR TITLE
[WIP] New Feature: `assume_role` connection config

### PR DESCRIPTION
**Pull request by brittandeyoung**
I am new to contributing to this project so some guidance to make sure I follow the correct procedures would be appreciated. This pull request is currently a Work In Progress. I am attempting to add the ability to use the `assume_role` in the connection configuration in order to leverage role assumption when configuring a connection to an AWS account instead of having to provide a `profile` or a static set of credentials. 

I did not see any integration tests related to the creating and testing of the AWS connection itself for me to run. 

Closes: #1844

Example config:

```
connection "aws" {
  plugin = "aws"
  type        = "aggregator"
  connections = ["aws_sandbox", "aws_ci", "aws_pre_prod"]
}

connection "aws_sandbox" {
  plugin      = "aws" 
  assume_role = {
    role_arn = "arn:aws:iam::111111111111:role/steampipe"
    duration = "300s"
    external_id = "testing"
    session_name = "steampipe"
  } 
  regions     = ["us-east-1", "us-west-2"]
}

```

I was able to use this config to run successful queries

### CODE REVIEW

**Comment by rajlearner17** [comment]
@brittandeyoung Thanks for trying out Steampipe!
We appreciate your interest in contributing to the community 👍 

While this will go through validation and further engagement from the respective members, please check the following docs for your reference

- [Writing Plugins](https://steampipe.io/docs/develop/writing-plugins)
- [AssumeRole](https://hub.steampipe.io/plugins/turbot/aws#assumerole-credentials-no-mfa) - I am not sure if this is the one you are looking for in your PR?

**Comment by brittandeyoung** [comment]
For awareness, I have the assume role configuration working. The last piece is the optional items under the `assume_role` config. More details are in the issue. With the current code in the PR, assuming roles works, but the optional fields `duration`, `external_id` and `session_name` do not work.

File: aws/service.go [hunk 1]
```diff
@@ -1708,7 +1709,7 @@ func getBaseClientForAccount(ctx context.Context, d *plugin.QueryData) (*aws.Con
 // If we expire the cache regularly we are causing SSO sessions to end
 // prematurely, and causing the AWS SDK to refresh credentials more often
 // using the IDMS service etc.
-var getBaseClientForAccountCached = plugin.HydrateFunc(getBaseClientForAccountUncached).Memoize(memoize.WithTtl(time.Hour * 24 * 30))
+var getBaseClientForAccountCached = plugin.HydrateFunc(getBaseClientForAccountUncached).Memoize(memoize.WithTtl(time.Second * 10))
```
Comment (cbruno10) [inline aws/service.go:1712]: @brittandeyoung Is there a reason you shortened the cache time, was this used just for testing?

File: config/aws.spc [hunk 2]
```diff
@@ -27,6 +27,14 @@ connection "aws" {
   # from an AWS credential file with the `profile` argument:
   #profile = "myprofile"
 
+  # If no credentials are specified, the plugin will use the AWS credentials
```
Comment (cbruno10) [inline config/aws.spc:30]: Is this example still correct, or does it follow the nested structure now?

File: aws/connection_config.go [hunk 3]
```diff
@@ -5,59 +5,28 @@ import (
...
-	Profile               *string  `cty:"profile"`
-	AccessKey             *string  `cty:"access_key"`
-	SecretKey             *string  `cty:"secret_key"`
-	SessionToken          *string  `cty:"session_token"`
-	MaxErrorRetryAttempts *int     `cty:"max_error_retry_attempts"`
-	MinErrorRetryDelay    *int     `cty:"min_error_retry_delay"`
-	IgnoreErrorCodes      []string `cty:"ignore_error_codes"`
-	EndpointUrl           *string  `cty:"endpoint_url"`
-	S3ForcePathStyle      *bool    `cty:"s3_force_path_style"`
+	Regions               []string          `hcl:"regions,optional"`
```
Comment (cbruno10) [inline aws/connection_config.go:11]: @brittandeyoung Referencing your most recent issue comment https://github.com/turbot/steampipe-plugin-aws/issues/1844#issuecomment-1722562384, you mention using `cty`, but here in the code I see all fields using `hcl` instead.

How does the code currently behave? Does it not work in some scenarios?

File: config/aws.spc (hunk 2 above)
Comment (brittandeyoung) [reply config/aws.spc:30]: @cbruno10 Thank you, I had forgot to update the example after changing to a map for configuration. The example has now been updated.

File: aws/connection_config.go (hunk 3 above)
Comment (brittandeyoung) [reply aws/connection_config.go:11]: @cbruno10 In the current state, the `role_arn` loads in properly and assuming the role works as expected as it has the `cty` tag. 

Currently the remainder of the fields to not have the `cty` tag. Without this tag, these items DO NOT load in from the config. This means these values are `nil`.

If I add the `cty` tag, the values do load in, but then are no longer optional and must be provided. If you do not provide them all, steampipe crashes and fails to load. If i attempt to add the `optional` tag, steampipe crashes when attempting to load the config. 

Any tips on direction from here is appreciated

**Comment by opsanatoliy** [comment]
@brittandeyoung could you please tell me how you replace aws plugin on your local version to run?

**Comment by opsanatoliy** [comment]
@brittandeyoung could you please tell me how you replace aws plugin on your local version to run?

**Comment by brittandeyoung** [comment]
> @brittandeyoung could you please tell me how you replace aws plugin on your local version to run?

@opsanatoliy Using the `install` make target. So:

```
make install
```

**Comment by opsanatoliy** [comment]
@brittandeyoung didn't you get anything like  `aws plugin failed to start ` when run `steampipe plugin list` 
I am trying to do something similar to what you already did in your PR, trying to figure out if it's a problem in code or in a config

File: aws/connection_config.go [hunk 4]
```diff
@@ -5,59 +5,28 @@ import (
...
+	SecretKey             *string           `hcl:"secret_key,optional"`
+	SessionToken          *string           `hcl:"session_token,optional"`
+	MaxErrorRetryAttempts *int              `hcl:"max_error_retry_attempts,optional"`
+	MinErrorRetryDelay    *int              `hcl:"min_error_retry_delay,optional"`
+	IgnoreErrorCodes      []string          `hcl:"ignore_error_codes,optional"`
+	EndpointUrl           *string           `hcl:"endpoint_url,optional"`
+	S3ForcePathStyle      *bool             `hcl:"s3_force_path_style,optional"`
 }
 
-var ConfigSchema = map[string]*schema.Attribute{
```
Comment (cbruno10) [inline aws/connection_config.go:25]: @brittandeyoung Why did you remove this along with `Schema:      ConfigSchema,` in plugin.go?

Comment (brittandeyoung) [reply aws/connection_config.go:25]: @cbruno10 This was done in order to support the map for `assume_role` as the schema does not support map objects. The idea was to use tagging to dynamically pull in the settings. 

I had this working with root level configurations instead of a map and utilizing the defined schema. Would this pull request be better to be reverted to using root level configurations instead of a map configuration?

Comment (bigdatasourav) [reply aws/connection_config.go:25]: Thanks for highlighting this; I have raised a Steampipe SDK [issue](https://github.com/turbot/steampipe-plugin-sdk/issues/691) to track.

File: aws/connection_config.go (hunk 3 above)
Comment (bigdatasourav) [reply aws/connection_config.go:11]: We were able to reproduce the above issue - 
```
Error: failed to start plugin 'hub.steampipe.io/plugins/turbot/aws@latest': failed to start 'hub.steampipe.io/plugins/turbot/aws@latest': failed to parse connection config for connection 'aws':
Unsuitable value type: Unsuitable value: attribute "external_id" is required
```

This could be an HCL parsing error - https://github.com/turbot/steampipe-plugin-sdk/blob/main/plugin/connection_config_schema.go#L88

I have raised a Steampipe SDK [issue](https://github.com/turbot/steampipe-plugin-sdk/issues/692) for the same with the details.

Comment (bigdatasourav) [reply aws/connection_config.go:11]: @brittandeyoung, please let us know if you are getting any other issues than the above.

File: aws/connection_config.go (hunk 4 above)
Comment (brittandeyoung) [reply aws/connection_config.go:25]: @bigdatasourav Great! This getting added to the SDK will likely enable me to complete this PR with working code.

File: aws/connection_config.go (hunk 3 above)
Comment (brittandeyoung) [reply aws/connection_config.go:11]: @bigdatasourav Great thanks! That was the only issue i was running into. I will keep an eye on this issue.

**Comment by misraved** [comment]
Closing the PR for now, since it is waiting on the resolution of the following SDK issues - 
https://github.com/turbot/steampipe-plugin-sdk/issues/692
https://github.com/turbot/steampipe-plugin-sdk/issues/691
